libbeat/common/mapstr.go

```diff
@@ -156,6 +156,38 @@ func (m MapStr) String() string {
 	return string(bytes)
 }
 
+// Flatten flattens the given MapStr and returns a flat MapStr
+//
+// Example:
+//   "hello": MapStr{"world": "test" }
+//
+// This is converted to:
+//   "hello.world": "test"
+//
+// This can be useful for testing or logging
+func (m MapStr) Flatten() MapStr {
+	return m.flatten("")
+}
+
+// flatten is a helper for Flatten. See docs Flatten
+func (m MapStr) flatten(prefix string) MapStr {
+	fields := MapStr{}
```

Comment:
Instead of creating a new MapStr every time this steps into an inner object, you could pass the output MapStr as an argument and have the function write directly to the final MapStr. This would avoid the call to `Update()`.